### PR TITLE
feat(manager/pipenv): Support custom environment variable usage in Pipfile source URLs 

### PR DESCRIPTION
**Pull request by fivetide**
## Changes

This change expands functionality of pipenv to allow for arbitrary credential placeholders as outlined in discussion #26447.
In addition, multiple source-urls can now be handled. 

## Context

- #26447 
- 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [X] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [X] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — DISMISSED** [review]
@fivetide thank you for this PR!

I found the logic flow a little hard to follow, and think this will be a little hard to maintain. I know you're mostly just extending the existing flow, but the additional complexity has now increased an existing problem of it being hard to understand.

Would you mind adding some comments to the flow so that it's easier to follow and for future maintainers? Sometimes improved variable names make commenting unnecessary too.

Once you're done with that, I'll see if there's anything I can add myself

File: lib/modules/manager/pipenv/artifacts.spec.ts [hunk 1]
```diff
@@ -12,19 +12,26 @@ import {
...
+  extractEnvironmentVariableName,
+} from './artifacts';
 import type { PipfileLockSchema } from './schema';
 import { updateArtifacts } from '.';
 
 const datasource = mocked(_datasource);
 const find = mockedFunction(_find);
 const getPkgReleases = mockedFunction(_getPkgReleases);
 
+jest.mock('../../../logger');
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.spec.ts:34]:
```suggestion
```
mocked globally

File: lib/modules/manager/pipenv/artifacts.spec.ts [hunk 2]
```diff
@@ -12,19 +12,26 @@ import {
...
 import { GlobalConfig } from '../../../config/global';
 import type { RepoGlobalConfig } from '../../../config/types';
+import { logger } from '../../../logger';
 import * as docker from '../../../util/exec/docker';
+import type { ExtraEnv, Opt } from '../../../util/exec/types';
 import type { StatusResult } from '../../../util/git/types';
 import { find as _find } from '../../../util/host-rules';
-import { getPkgReleases as _getPkgReleases } from '../../datasource';
 import * as _datasource from '../../datasource';
+import { getPkgReleases as _getPkgReleases } from '../../datasource';
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.spec.ts:21]: this reorder is probably the cause of failing ci.

wonder why it's imported twice 🤔

Comment (fivetide) [reply lib/modules/manager/pipenv/artifacts.spec.ts:20]: no clue, maybe intellij found it a great idea

File: lib/modules/manager/pipenv/artifacts.ts [hunk 3]
```diff
@@ -113,35 +114,95 @@ export function getPipenvConstraint(
...
+  const sourceUrls = await findPipfileSourceUrlsWithCredentials(
+    newPipfileContent,
+    pipfileName,
+  );
+  for (const sourceUrl of sourceUrls) {
+    const parsedSourceUrl = parseUrl(sourceUrl);
+    if (!parsedSourceUrl) {
+      logger.warn(`Pipenv: could not parse url for source: ${sourceUrl}`);
+      return;
+    }
```
Comment (rarkins) [inline lib/modules/manager/pipenv/artifacts.ts:192]: In theory this should never be reached, because the list of sourceUrls was already filtered using a parseUrl?

File: lib/modules/manager/pipenv/artifacts.ts [hunk 4]
```diff
@@ -113,35 +114,95 @@ export function getPipenvConstraint(
...
+    logger.debug(`Trying to add credentials for ${sourceUrl}`);
+    const { password, username } = getMatchingHostRule(sourceUrl) ?? {};
+    if (username) {
+      const environmentVariableName =
+        extractEnvironmentVariableName(parsedSourceUrl.username) ?? 'USERNAME';
```
Comment (rarkins) [inline lib/modules/manager/pipenv/artifacts.ts:197]: Are these fallbacks to USERNAME and PASSWORD needed in any known scenarios or there "just in case"?

File: lib/modules/manager/pipenv/artifacts.ts [hunk 5]
```diff
@@ -113,35 +114,95 @@ export function getPipenvConstraint(
...
-  return find({ hostType: PypiDatasource.id, url });
+function getMatchingHostRule(url: string): HostRule | undefined {
+  const parsedUrl = parseUrl(url);
+  if (parsedUrl) {
+    parsedUrl.username = '';
+    parsedUrl.password = '';
+    const urlWithoutCredentials = parsedUrl.toString();
+
+    return find({ hostType: PypiDatasource.id, url: urlWithoutCredentials });
+  }
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.ts:125]:
```suggestion
  }
  return undefined;
```
missing return?

File: lib/modules/manager/pipenv/artifacts.ts [hunk 6]
```diff
@@ -113,35 +114,95 @@ export function getPipenvConstraint(
...
   pipfileContent: string,
   pipfileName: string,
-): Promise<string | null> {
+): Promise<string[]> {
   const pipfile = await extractPackageFile(pipfileContent, pipfileName);
-  if (!pipfile) {
-    logger.debug('Error parsing Pipfile');
+
+  return (
+    pipfile?.registryUrls?.filter((url) => !!parseUrl(url)?.username) ?? []
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.ts:135]: please return the parsed urls, so we don't need to parse multiple times (perf)

File: lib/modules/manager/pipenv/artifacts.ts (hunk 3 above)
Comment (fivetide) [reply lib/modules/manager/pipenv/artifacts.ts:192]: correct. tsc just did not understand. that branch will be gone after next commit.

File: lib/modules/manager/pipenv/artifacts.ts (hunk 4 above)
Comment (fivetide) [reply lib/modules/manager/pipenv/artifacts.ts:197]: initially as fallback, but now they are not needed anymore. will remove

File: lib/modules/manager/pipenv/artifacts.ts (hunk 5 above)
Comment (fivetide) [reply lib/modules/manager/pipenv/artifacts.ts:125]: omitted intentionally. i didnt want to introduce a reason for the CI to demand coverage for an explicit statement. But lets see if it does...

File: lib/modules/manager/pipenv/artifacts.spec.ts [hunk 7]
```diff
@@ -1,3 +1,4 @@
+import { expect } from '@jest/globals';
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.spec.ts:1]:
```suggestion
```

File: lib/modules/manager/pipenv/artifacts.spec.ts (hunk 2 above)
Comment (viceice) [reply lib/modules/manager/pipenv/artifacts.spec.ts:22]: - #28337

File: lib/modules/manager/pipenv/artifacts.ts [hunk 8]
```diff
@@ -113,35 +114,100 @@ export function getPipenvConstraint(
...
-    '$USERNAME:',
-    // eslint-disable-next-line no-template-curly-in-string
-    '${USERNAME}',
-    '$PASSWORD@',
-    // eslint-disable-next-line no-template-curly-in-string
-    '${PASSWORD}',
-  ];
+  return (pipfile?.registryUrls
+    ?.map(parseUrl)
+    ?.filter((url) => !!url?.username) ?? []) as URL[]; //tsc does not notice null entries are removed
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.ts:137]:
```suggestion
  return pipfile?.registryUrls
    .filter(is.urlInstance)
    .filter((url) => is.nonEmptyStringAndNotWhitespace(url.username)) ?? [];
```

File: lib/modules/manager/pipenv/artifacts.ts [hunk 9]
```diff
@@ -113,35 +114,100 @@ export function getPipenvConstraint(
...
+}
+
+/**
+ * This will extract the actual variable name from an environment-placeholder:
+ * ${USERNAME:-defaultvalue} will yield 'USERNAME'
+ */
+export function extractEnvironmentVariableName(
+  credential: string,
+): string | null {
+  const match = decodeURI(credential).match('([a-zA-Z0-9_]+)');
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.ts:147]:
```suggestion
  const match = regEx('([a-z0-9_]+)', 'i').exec(decodeURI(credential));
```
(perf)

File: lib/modules/manager/pipenv/artifacts.ts [hunk 10]
```diff
@@ -113,35 +114,100 @@ export function getPipenvConstraint(
...
+}
 
-  const sourceWithCredentials = pipfile.registryUrls?.find((url) =>
-    credentialTokens.some((token) => url.includes(token)),
+export function addExtraEnvVariable(
+  extraEnv: ExtraEnv<unknown>,
+  environmentVariableName: string,
+  environmentValue: string,
+): void {
+  logger.debug(
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.ts:156]:
```suggestion
  logger.trace(
```

you can remap via config if required

https://docs.renovatebot.com/configuration-options/#loglevelremap

File: lib/modules/manager/pipenv/artifacts.ts [hunk 11]
```diff
@@ -113,35 +114,100 @@ export function getPipenvConstraint(
...
+  newPipfileContent: string,
+  pipfileName: string,
+  extraEnv: ExtraEnv<unknown>,
+): Promise<void> {
+  const sourceUrls = await findPipfileSourceUrlsWithCredentials(
+    newPipfileContent,
+    pipfileName,
+  );
+  for (const parsedSourceUrl of sourceUrls) {
+    logger.debug(`Trying to add credentials for ${parsedSourceUrl.toString()}`);
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.ts:186]:
```suggestion
    logger.trace(`Trying to add credentials for ${parsedSourceUrl}`);
```
Do we need the `tostring`?

File: lib/modules/manager/pipenv/artifacts.spec.ts (hunk 2 above)
Comment (viceice) [reply lib/modules/manager/pipenv/artifacts.spec.ts:21]: please merge main to fix this

Comment (fivetide) [reply lib/modules/manager/pipenv/artifacts.spec.ts:21]: done :)

File: lib/modules/manager/pipenv/artifacts.spec.ts [hunk 12]
```diff
@@ -626,7 +633,34 @@ describe('modules/manager/pipenv/artifacts', () => {
...
+      );
+    }
+  });
+
+  it('warns about duplicate placeholders with different values', () => {
+    const extraEnv: Opt<ExtraEnv> = {
+      FOO: '1',
+    };
+    addExtraEnvVariable(extraEnv, 'FOO', '2');
+    expect(logger.warn).toHaveBeenCalledTimes(1);
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.spec.ts:660]:
```suggestion
    expect(logger.warn).toHaveBeenCalledOnce();
```

File: lib/modules/manager/pipenv/artifacts.spec.ts [hunk 13]
```diff
@@ -626,7 +633,34 @@ describe('modules/manager/pipenv/artifacts', () => {
     ]);
   });
 
-  it('does not pass private credential environment vars if variable names differ from allowed', async () => {
+  it('returns no host rule on invalid url', () => {
+    expect(getMatchingHostRule('')).toBeNull();
+  });
+
+  it('extracts correct environment variable from credential placeholder', () => {
+    for (const testCase of [
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.spec.ts:641]: use jest it.each

File: lib/modules/manager/pipenv/artifacts.spec.ts [hunk 14]
```diff
@@ -637,27 +637,28 @@ describe('modules/manager/pipenv/artifacts', () => {
...
-  it('extracts correct environment variable from credential placeholder', () => {
-    for (const testCase of [
-      { credential: '$USERNAME', result: 'USERNAME' },
-      { credential: '$', result: null },
-      { credential: '', result: null },
-      { credential: '${USERNAME}', result: 'USERNAME' },
-      { credential: '${USERNAME:-default}', result: 'USERNAME' },
-      { credential: '${COMPLEX_NAME_1:-default}', result: 'COMPLEX_NAME_1' },
-    ]) {
+  it.each([
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.spec.ts:640]: can you please use the tagged template syntax instead?

- https://github.com/renovatebot/renovate/blob/main/docs/development/best-practices.md#unit-testing

File: lib/modules/manager/pipenv/artifacts.spec.ts [hunk 15]
```diff
@@ -626,7 +633,34 @@ describe('modules/manager/pipenv/artifacts', () => {
...
+  it.each`
+    credential                      | result
+    ${'$USERNAME'}                  | ${'USERNAME'}
+    ${'$'}                          | ${null}
+    ${''}                           | ${null}
+    ${'${USERNAME}'}                | ${'USERNAME'}
+    ${'${USERNAME:-default}'}       | ${'USERNAME'}
+    ${'${COMPLEX_NAME_1:-default}'} | ${'COMPLEX_NAME_1'}
+  `(
+    'extracts correct environment variable from credential placeholder',
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.spec.ts:649]:
```suggestion
    'extracts correct environment variable from credential placeholder for `$credential`',
```

**Comment by rarkins** [comment]
@fivetide it's best that you don't merge from main unless it's to fix something. Otherwise it resets the tests and I need to manually approve it again. If/when @viceice it will start automerging, assuming tests have passed

**Comment by fivetide** [comment]
Hey @rarkins, @viceice  thanks for getting this PR ready. It looks to me that the merge is blocked by outdated changes. can you point me to how to resolve that?

**Comment by rarkins** [comment]
@fivetide it will now automerge once Michael approves

**Review by viceice — APPROVED** [review]

**Comment by fivetide** [comment]
Thanks a whole lot to you guys for helping me making this happen!
